File: articles/nature-nurture-and-the-developmental-question/python/model_nature_nurture_development.py
#!/usr/bin/env python3
"""Model synthetic nature-nurture development and differential susceptibility."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "nature_nurture_development_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_nature_nurture_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    dynamic_model = smf.ols(
        """
        development_score ~ lag_score + time + biological_sensitivity +
        caregiver_support + acute_stress + structural_risk + chronic_adversity +
        institutional_support + disability_support + resource_stability +
        intervention + protective_context +
        biological_sensitivity:caregiver_support +
        biological_sensitivity:acute_stress +
        biological_sensitivity:protective_context
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    profile_model = smf.ols(
        """
        development_score ~ time + biological_sensitivity + structural_risk +
        chronic_adversity + protective_context + acute_stress +
        C(sensitivity_profile)
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_nature_nurture_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC NATURE-NURTURE DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(dynamic_model.summary().as_text())
        f.write("\n\nSENSITIVITY PROFILE MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(profile_model.summary().as_text())

    trajectory = panel.groupby(["time", "structural_risk"], as_index=False).agg(
        average_development=("development_score", "mean"),
        average_support=("caregiver_support", "mean"),
        average_stress=("acute_stress", "mean"),
        average_protective_context=("protective_context", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["risk_group"] = trajectory["structural_risk"].map({
        0: "Lower structural risk",
        1: "Higher structural risk",
    })
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_structural_risk_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("risk_group"):
        plt.plot(group["time"], group["average_development"], marker="o", label=str(group_name))
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Developmental Trajectories by Structural Risk")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_structural_risk_trajectory.png", dpi=160)
    plt.close()

    protective = panel.groupby("time", as_index=False).agg(
        average_caregiver_support=("caregiver_support", "mean"),
        average_acute_stress=("acute_stress", "mean"),
        average_protective_context=("protective_context", "mean"),
        average_intervention=("intervention", "mean"),
        average_development=("development_score", "mean"),
    )
    protective.to_csv(OUTPUTS_DIR / "python_protective_context_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(protective["time"], protective["average_caregiver_support"], marker="o", label="caregiver support")
    plt.plot(protective["time"], protective["average_acute_stress"], marker="o", label="acute stress")
    plt.plot(protective["time"], protective["average_protective_context"], marker="o", label="protective context")
    plt.plot(protective["time"], protective["average_intervention"], marker="o", label="intervention")
    plt.xlabel("Time")
    plt.ylabel("Average index")
    plt.title("Synthetic Nature-Nurture Protective Context Over Time")
    plt.legend(fontsize=8)
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_protective_context_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("sensitivity_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_development=("development_score", "mean"),
        average_protective_context=("protective_context", "mean"),
        average_stress=("acute_stress", "mean"),
        average_sensitivity=("biological_sensitivity", "mean"),
        average_structural_risk=("structural_risk", "mean"),
        average_chronic_adversity=("chronic_adversity", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_sensitivity_profiles.csv", index=False)

    school_summary = panel.groupby("school_id", as_index=False).agg(
        institutional_support=("institutional_support", "mean"),
        disability_support=("disability_support", "mean"),
        resource_stability=("resource_stability", "mean"),
        average_development=("development_score", "mean"),
        average_stress=("acute_stress", "mean"),
        average_protective_context=("protective_context", "mean"),
    )
    school_summary.to_csv(OUTPUTS_DIR / "python_school_context_summary.csv", index=False)

    print("Wrote Python nature-nurture model outputs.")


if __name__ == "__main__":
    main()

File: articles/nature-nurture-and-the-developmental-question/r/model_nature_nurture_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "nature_nurture_development_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_nature_nurture_panel.py first.")

panel <- read.csv(data_path)

model <- lmer(
  development_score ~ time + biological_sensitivity + caregiver_support +
    acute_stress + structural_risk + chronic_adversity +
    institutional_support + disability_support + resource_stability +
    intervention + protective_context +
    biological_sensitivity:caregiver_support +
    biological_sensitivity:acute_stress +
    biological_sensitivity:protective_context +
    (1 + time | school_id/child_id),
  data = panel
)

capture.output(
  summary(model),
  file = file.path(outputs_dir, "r_nature_nurture_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, structural_risk) |>
  summarize(
    average_development = mean(development_score),
    average_support = mean(caregiver_support),
    average_stress = mean(acute_stress),
    average_protective_context = mean(protective_context),
    standard_error = sd(development_score) / sqrt(n()),
    .groups = "drop"
  ) |>
  mutate(
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    risk_group = ifelse(structural_risk == 1, "Higher structural risk", "Lower structural risk")
  )

write.csv(trajectory, file.path(outputs_dir, "r_structural_risk_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_structural_risk_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development, linetype = risk_group)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = risk_group), alpha = 0.12) +
    labs(title = "Synthetic Developmental Trajectories by Structural Risk", x = "Time", y = "Average development score", linetype = "Group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

support_summary <- panel |>
  group_by(time) |>
  summarize(
    average_caregiver_support = mean(caregiver_support),
    average_acute_stress = mean(acute_stress),
    average_protective_context = mean(protective_context),
    average_intervention = mean(intervention),
    average_development = mean(development_score),
    .groups = "drop"
  )

write.csv(support_summary, file.path(outputs_dir, "r_protective_context_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_protective_context_trajectory.png"),
  plot = ggplot(support_summary, aes(x = time)) +
    geom_line(aes(y = average_caregiver_support, linetype = "caregiver support"), linewidth = 1) +
    geom_line(aes(y = average_acute_stress, linetype = "acute stress"), linewidth = 1) +
    geom_line(aes(y = average_protective_context, linetype = "protective context"), linewidth = 1) +
    geom_line(aes(y = average_intervention, linetype = "intervention"), linewidth = 1) +
    labs(title = "Synthetic Nature-Nurture Protective Context Over Time", x = "Time", y = "Average index", linetype = "Measure") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R nature-nurture development outputs.")
